### PR TITLE
Parallelise package extraction

### LINKED ISSUE #6: Analyze packages in parallel
Instead of running packages one by one, the process can be parallelized.

### PR DESCRIPTION
**Pull request by Vel1khan**
Closes #6 

I'm seeing speedups of 4-5x on my machine. Only thing that is missing is integration with #16. This can be done after that issue is completed.

### CODE REVIEW

File: analyzer/src/main/java/nl/tudelft/Runner.java [hunk 1]
```diff
@@ -29,30 +34,44 @@ Runner addExtractor(Extractor extractor) throws SQLException {
...
-                fetchEnd = Instant.now();
-                values = extractInto(mvn, pkg);
-            } catch (PackageException e) {
-                LOGGER.error(e);
-                continue;
-            }
+    private void processPackages(Collection<PackageId> packages, Field[] fields, Maven mvn) {
+        int cores = Runtime.getRuntime().availableProcessors();
+        LOGGER.info("Number of cores available = " + cores);
+        ExecutorService executor = Executors.newFixedThreadPool(cores);
```
Comment (St33lPhoenix) [inline analyzer/src/main/java/nl/tudelft/Runner.java]: This is nonsensical, the number of physical cores is unrelated to software threading. This should just be some number configurable in `config.yml`.

File: analyzer/src/main/java/nl/tudelft/Runner.java [hunk 2]
```diff
@@ -29,30 +34,44 @@ Runner addExtractor(Extractor extractor) throws SQLException {
...
+        // Combine all CompletableFutures into a single CompletableFuture then .get() to wait for all threads
+        // to complete (successfully or exceptionally)
+        try {
+            CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).get();
+        } catch (InterruptedException e) {
+            LOGGER.error(e);
+        } catch (ExecutionException e) {
+            // No need to log here. We already log inside each thread,
+            // this exception will contain only the exception thrown by
+            // the first Future to throw an exception.
```
Comment (St33lPhoenix) [inline analyzer/src/main/java/nl/tudelft/Runner.java]: There can still be unexpected exceptions, so they should not be ignored.

File: analyzer/src/main/java/nl/tudelft/Runner.java [hunk 3]
```diff
@@ -29,30 +34,44 @@ Runner addExtractor(Extractor extractor) throws SQLException {
...
+            executor.submit(new ProcessPackageTask(id, fields, mvn, future));
+            futures.add(future);
+        }
+
+        // Combine all CompletableFutures into a single CompletableFuture then .get() to wait for all threads
+        // to complete (successfully or exceptionally)
+        try {
+            CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).get();
+        } catch (InterruptedException e) {
+            LOGGER.error(e);
```
Comment (St33lPhoenix) [inline analyzer/src/main/java/nl/tudelft/Runner.java:69]: This should be propagated, i.e. `Thread.currentThread().interrupt()`

File: analyzer/src/main/java/nl/tudelft/Runner.java [hunk 4]
```diff
@@ -73,4 +92,64 @@ private List<Object> extractInto(Maven mvn, Package pkg) throws IOException {
     public void close() throws IOException {
         db.close();
     }
+
+    /**
+     *
+     */
+    private class ProcessPackageTask implements Callable{
```
Comment (St33lPhoenix) [inline analyzer/src/main/java/nl/tudelft/Runner.java]: As far as I'm aware, Callable is a generic interface, so why is its type not defined?

File: analyzer/src/main/java/nl/tudelft/Runner.java [hunk 5]
```diff
@@ -73,4 +92,64 @@ private List<Object> extractInto(Maven mvn, Package pkg) throws IOException {
...
+         * @return 
+         * @throws SQLException
+         */
+        @Override
+        public Object call() {
+            if (cancelled.get()) {
+                LOGGER.error("SQL Exception encountered in another thread. Skipping package " + id);
+                future.completeExceptionally(new SQLException("Lost connection to DB!"));
+                return null;
+            }
```
Comment (St33lPhoenix) [inline analyzer/src/main/java/nl/tudelft/Runner.java:120]: Why would every task need to be cancelled if one runner fails? (and these errors do not necessarily indicate a loss of connection to the db, in which case one should just attempt to reconnect).

File: analyzer/src/main/java/nl/tudelft/Runner.java [hunk 6]
```diff
@@ -73,4 +92,64 @@ private List<Object> extractInto(Maven mvn, Package pkg) throws IOException {
...
+            this.id = id;
+            this.fields = fields;
+            this.mvn = mvn;
+            this.future = future;
+        }
+
+        /**
+         * @return 
+         * @throws SQLException
+         */
```
Comment (St33lPhoenix) [inline analyzer/src/main/java/nl/tudelft/Runner.java]: Incomplete / invalid docs?

File: analyzer/src/main/java/nl/tudelft/Runner.java [hunk 7]
```diff
@@ -73,4 +92,64 @@ private List<Object> extractInto(Maven mvn, Package pkg) throws IOException {
...
+            } catch (PackageException | IOException e) {
+                LOGGER.error(e);
+                future.complete(null);
+                // TODO Put this package in the unresolved table
+                return null;
+            }
+
+            try {
+                db.update(id, fields, values.toArray());
+            } catch (SQLException e) {
```
Comment (St33lPhoenix) [inline analyzer/src/main/java/nl/tudelft/Runner.java:138]: Why is this not logged?

File: analyzer/src/main/java/nl/tudelft/Runner.java (hunk 1 above)
Comment (Vel1khan) [reply analyzer/src/main/java/nl/tudelft/Runner.java]: Made it ocnfigurable

File: analyzer/src/main/java/nl/tudelft/Runner.java (hunk 5 above)
Comment (Vel1khan) [reply analyzer/src/main/java/nl/tudelft/Runner.java:120]: Continue in another issue

**Pull request merged**